compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1613,42 +1613,65 @@ impl<'a> Parser<'a> {
                 Applicability::HasPlaceholders,
             );
             return Some(ident);
-        } else if let PatKind::Ident(_, ident, _) = pat.kind {
-            if require_name
-                && (self.token == token::Comma
-                    || self.token == token::Lt
-                    || self.token == token::CloseDelim(token::Paren))
-            {
-                // `fn foo(a, b) {}`, `fn foo(a<x>, b<y>) {}` or `fn foo(usize, usize) {}`
-                if first_param {
-                    err.span_suggestion(
-                        pat.span,
-                        "if this is a `self` type, give it a parameter name",
-                        format!("self: {}", ident),
-                        Applicability::MaybeIncorrect,
-                    );
-                }
-                // Avoid suggesting that `fn foo(HashMap<u32>)` is fixed with a change to
-                // `fn foo(HashMap: TypeName<u32>)`.
-                if self.token != token::Lt {
-                    err.span_suggestion(
-                        pat.span,
-                        "if this is a parameter name, give it a type",
-                        format!("{}: TypeName", ident),
-                        Applicability::HasPlaceholders,
-                    );
+        } else if require_name
+            && (self.token == token::Comma
+                || self.token == token::Lt
+                || self.token == token::CloseDelim(token::Paren))
+        {
+            let (ident, self_sugg, param_sugg, type_sugg) = match pat.kind {
+                PatKind::Ident(_, ident, _) => (
+                    ident,
+                    format!("self: {}", ident),
+                    format!("{}: TypeName", ident),
+                    format!("_: {}", ident),
+                ),
+                // Also catches `fn foo(&a)`.
+                PatKind::Ref(ref pat, mutab) => {
+                    if let PatKind::Ident(_, ident, _) = pat.clone().into_inner().kind {
+                        let mutab = mutab.prefix_str();
+                        (
+                            ident,
+                            format!("self: &{}{}", mutab, ident),
+                            format!("{}: &{}TypeName", ident, mutab),
+                            format!("_: &{}{}", mutab, ident),
+                        )
+                    } else {
```

Comment:
@JohnTitor can you also add a test case for a fully qualified path, like in the original report? I get the feeling that the new code will work well _only_ for the case of `fn foo(&bar)` and not for `fn foo(&<Bar as T>::Baz)`. To handle the later, we might want to add a Parser snapshot here and try to parse a type. If successful, we know we can suggest `_: <Ty>` and recover the parser state.

Reply:
Good point! Yeah, the current `parameter_without_type` doesn't handle it with/without refs (1082ced). Should we address that case in this PR?

Reply:
You can see an example of what I mean here:

https://github.com/rust-lang/rust/blob/6f43f4c3fbc99c3a6699ed8d44608c8f086b10af/compiler/rustc_parse/src/parser/diagnostics.rs#L749-L761

You have a bounded lookahead to see that you might be in a weird known situation that will for sure emit an error. You clone the parser and try an alternative parse. If it is successful, you emit a targeted error, otherwise you "revert" the parser by overriding the parser state with your snapshot. Doing this you gain a lot of confidence in any suggestions you give because you _know_ that the alternative parse worked.